packages/instruction-plans/src/instruction-plan.ts

```diff
@@ -1 +1,526 @@
-export type InstructionPlan = unknown;
+import { Instruction } from '@solana/instructions';
+import {
+    appendTransactionMessageInstruction,
+    BaseTransactionMessage,
+    TransactionMessageWithFeePayer,
+} from '@solana/transaction-messages';
+import { getTransactionMessageSize, TRANSACTION_SIZE_LIMIT } from '@solana/transactions';
+
+/**
+ * A set of instructions with constraints on how they can be executed.
+ *
+ * This is structured as a recursive tree of plans in order to allow for
+ * parallel execution, sequential execution and combinations of both.
+ *
+ * Namely the following plans are supported:
+ * - {@link SingleInstructionPlan} - A plan that contains a single instruction.
+ *   This is a simple instruction wrapper and the simplest leaf in this tree.
+ * - {@link ParallelInstructionPlan} - A plan that contains other plans that
+ *   can be executed in parallel.
+ * - {@link SequentialInstructionPlan} - A plan that contains other plans that
+ *   must be executed sequentially. It also defines whether the plan is divisible
+ *   meaning that instructions inside it can be split into separate transactions.
+ * - {@link MessagePackerInstructionPlan} - A plan that can dynamically fill
+ *  instructions into transaction messages.
```

Comment:
Pack instructions? No big deal, I know updating PRs is a pain.
```suggestion
 * - {@link MessagePackerInstructionPlan} - A plan that can dynamically pack
 *   instructions into transaction messages.
```